pkg/operator/ceph/file/mirror/mirror.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package mirror for mirroring
+package mirror
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/banzaicloud/k8s-objectmatcher/patch"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/ceph/cluster/mon"
+	"github.com/rook/rook/pkg/operator/ceph/config"
+	"github.com/rook/rook/pkg/operator/ceph/controller"
+	opcontroller "github.com/rook/rook/pkg/operator/ceph/controller"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// AppName is the ceph filesystem mirror application name
+	AppName = "rook-ceph-fs-mirror"
+	// minimum amount of memory in MB to run the pod
+	cephFilesystemMirrorPodMinimumMemory uint64 = 512
+)
+
+var updateDeploymentAndWait = mon.UpdateCephDeploymentAndWait
+
+// Start begins the process of running filesystem mirroring daemons.
+func (r *ReconcileFilesystemMirror) start(filesystemMirror *cephv1.CephFilesystemMirror) error {
+	ctx := context.TODO()
+	// Validate pod's memory if specified
+	err := controller.CheckPodMemory(cephv1.ResourcesKeyFilesystemMirror, filesystemMirror.Spec.Resources, cephFilesystemMirrorPodMinimumMemory)
+	if err != nil {
+		return errors.Wrap(err, "error checking pod memory")
+	}
+
+	// Create the controller owner ref
+	// It will be associated to all resources of the CephRBDMirror
+	ref, err := opcontroller.GetControllerObjectOwnerReference(filesystemMirror, r.scheme)
+	if err != nil || ref == nil {
+		return errors.Wrapf(err, "failed to get controller %q owner reference", filesystemMirror.Name)
+	}
+
+	daemonID := k8sutil.IndexToName(0)
```

Comment:
Only a single mirroring configuration is allowed per rook cluster, right? Do we need to prevent multiple mirroring CRs from being created? (similar to how we only allow a single CephCluster CR in a namespace)

Reply:
> Only a single mirroring configuration is allowed per rook cluster, right? 

No, it's per filesystem.

> Do we need to prevent multiple mirroring CRs from being created? (similar to how we only allow a single CephCluster CR in a namespace)

Not really since we support multiple fs.

Reply:
Per discussion, agreed that only a single mirroring CR is supported, but no need for extra restrictions since we expect users to follow the docs and examples.